### PR TITLE
[SWY-175] Add chord parsing and transposition helpers

### PR DESCRIPTION
**Pull request by justaddcl**
## Background
- Linear: https://linear.app/sanbi/issue/SWY-175/add-pure-chord-parsing-and-transposition-helpers
- Adds pure chord parsing and transposition helpers so Sanbi can safely recognize chord tokens, calculate semitone movement between supported song keys, and transpose chord roots without coupling this logic to UI or data access.
- The helpers return structured status and review metadata for unsupported or ambiguous chord-like tokens, which gives future product flows a way to preserve content while surfacing cases that need human attention.

## Summary by CodeRabbit

* **New Features**
  * Added music chord parsing and transposition support, including handling for sharps, flats, slash chords, and key changes.
  * Exposed music utilities through a simpler top-level entry point for broader app access.

* **Bug Fixes**
  * Improved validation so unsupported chord inputs and song keys are handled safely.
  * Added review flags for ambiguous chord cases to help catch potentially incorrect transpositions.

* **Tests**
  * Added coverage for key conversion, chord parsing, transposition behavior, and unsupported input handling.

- Added `src/lib/music/chords` exports for song key validation, song key semitone lookup, transpose distance calculation, chord token parsing, chord root transposition, and full token transposition.
- Added Tonal note and interval dependencies to handle pitch-class and interval math instead of maintaining custom chromatic tables.
- Added flat-key spelling behavior for flat target keys, slash chord support, accidental normalization for ASCII and Unicode accidentals, and explicit result metadata for unsupported keys/tokens and review-required tokens.
- Added focused Jest coverage for supported keys, parser behavior, root transposition, semitone distance, and complete chord-token transposition outcomes.

## How to test
- In a future UI that wires these helpers into song content, set a song from one supported key to another and verify chord tokens transpose while non-chord lyric text remains unchanged.
- Check slash chords, flat target keys, sharp target keys, lowercase chord-like tokens, and bare root tokens to confirm ambiguous values are preserved with review indicators rather than silently discarded.

<h3>Greptile Summary</h3>

This PR introduces a self-contained chord parsing and transposition layer (`src/lib/music/chords`) backed by `@tonaljs/note` and `@tonaljs/interval`, replacing any future need for hand-rolled chromatic tables. All three issues surfaced in the previous review round (unreachable `undefined` guard in `normalizeChordAccidental`, duplicated `modulo` helper, and `wasTransposed` incorrectly flagging casing normalization as pitch movement) have been cleanly resolved.

- **New helpers**: `isSupportedSongKey`, `getSongKeySemitone`, `getTransposeSemitoneDistance`, `parseChordToken`/`parseChordPart`, `transposeChordRoot`, and `transposeChordToken` — each narrowly scoped and exported through a single `src/lib/music/chords` barrel.
- **Result metadata design**: `transposeChordToken` returns structured `status`, `wasTransposed`, `requiresReview`, and `reviewReasons` fields so future callers can preserve ambiguous tokens (bare-root, lowercase-root) and surface them for human review rather than silently dropping or mis-transposing them.
- **Flat/sharp spelling**: `FLAT_KEY_ROOTS` drives the pitch-class spelling decision, producing `Db`/`Eb`/`Ab` etc. for flat target keys and `C#`/`F#`/`G#` etc. for sharp target keys.

<h3>Confidence Score: 5/5</h3>

Safe to merge — the new helpers are pure functions with no side effects, all previous review threads are resolved, and the TonalJS delegates are used correctly throughout.

All logic is well-isolated, thoroughly tested for the happy path and key edge cases, and carries no data-access or UI coupling. The only gap is a missing test for the enharmonic-respelling path (e.g., C# → Db when the keys are enharmonically equivalent), which leaves the wasTransposed: false / differing transposedToken behaviour undocumented for future callers — worth adding before those callers are written, but not a blocker.

The transposeChordToken.test.ts file is missing a test case for enharmonic respelling (e.g., sourceKey: "c_sharp", targetKey: "d_flat"). All other files look correct.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/lib/music/chords/transposeChordToken.ts | Main entry point for chord transposition; correctly gates writes on both semitone distance and token equality, addressing previous review feedback on the wasTransposed field. |
| src/lib/music/chords/__tests__/transposeChordToken.test.ts | Covers the main transposition paths and the post-review wasTransposed fix, but lacks a test for enharmonic respelling (e.g., c_sharp → d_flat) where wasTransposed=false despite the token changing. |
| src/lib/music/chords/transposeChordRoot.ts | Correctly delegates pitch-class arithmetic to TonalJS and selects flat/sharp spelling via FLAT_KEY_ROOTS; modulo helper now shared with getTransposeSemitoneDistance. |
| src/lib/music/chords/parseChordPart.ts | CHORD_PART_PATTERN and CHORD_SUFFIX_PATTERN cover expected chord vocabulary; normalizes Unicode accidentals and validates the root against TonalJS before accepting. |
| src/lib/music/chords/getSongKeySemitone.ts | Correctly converts SongKey enum values to TonalJS note names and delegates semitone lookup to Note.get().chroma; spellSongKeyRoot is reused by other helpers. |
| src/lib/music/chords/getTransposeSemitoneDistance.ts | Computes the forward semitone distance between two song keys via TonalJS interval arithmetic and normalises to [0, 11] with the shared modulo helper. |
| src/lib/music/chords/types.ts | Clean type definitions; ChordRoot is a branded union that prevents invalid roots from silently flowing through the system. |
| src/lib/music/chords/constants/flatKeyRoots.ts | Enumerates the six standard flat major key roots; used to decide flat vs sharp spelling for transposed chord output. |
| src/lib/music/chords/normalizeChordAccidental.ts | Now correctly accepts string | undefined to align the parameter type with the runtime guard, addressing the previous review thread. |
| src/lib/music/chords/modulo.ts | Extracted shared positive-modulo helper, resolving the previous review comment about duplication. |
| src/lib/music/chords/parseChordToken.ts | Rejects tokens with leading/trailing whitespace, multiple slash parts, and empty bass segments; delegates per-part validation to parseChordPart. |

<h3>Flowchart</h3>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["transposeChordToken(token, sourceKey, targetKey)"] --> B{isSupportedSongKey\nsourceKey AND targetKey?}
    B -- No --> C["return createUnchangedResult\nstatus: 'unsupported-key'"]
    B -- Yes --> D["parseChordToken(token)"]
    D -- null --> E["return createUnchangedResult\nstatus: 'unsupported-token'"]
    D -- ParsedChordToken --> F["getTransposeSemitoneDistance\nsourceKey → targetKey\n[0..11]"]
    F --> G["transposeChordRoot\nroot + semitoneDistance\n+ getTargetSpelling"]
    G --> H{slash chord?}
    H -- Yes --> I["transposeChordRoot\nbass root"]
    H -- No --> J
    I --> J["assemble transposedToken"]
    J --> K["getReviewReasons\n(lowercase-root, bare-root)"]
    K --> L["return ChordTokenTranspositionResult\nwasTransposed = semitoneDistance≠0 AND token changed"]

    subgraph spellDecision ["getTargetSpelling"]
        M["spellSongKeyRoot(targetKey)"] --> N{root in\nFLAT_KEY_ROOTS?}
        N -- Yes --> O["spelling: flat\nNote.fromMidi"]
        N -- No --> P["spelling: sharp\nNote.fromMidiSharps"]
    end
    G -.-> spellDecision
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A["transposeChordToken(token, sourceKey, targetKey)"] --> B{isSupportedSongKey\nsourceKey AND targetKey?}
    B -- No --> C["return createUnchangedResult\nstatus: 'unsupported-key'"]
    B -- Yes --> D["parseChordToken(token)"]
    D -- null --> E["return createUnchangedResult\nstatus: 'unsupported-token'"]
    D -- ParsedChordToken --> F["getTransposeSemitoneDistance\nsourceKey → targetKey\n[0..11]"]
    F --> G["transposeChordRoot\nroot + semitoneDistance\n+ getTargetSpelling"]
    G --> H{slash chord?}
    H -- Yes --> I["transposeChordRoot\nbass root"]
    H -- No --> J
    I --> J["assemble transposedToken"]
    J --> K["getReviewReasons\n(lowercase-root, bare-root)"]
    K --> L["return ChordTokenTranspositionResult\nwasTransposed = semitoneDistance≠0 AND token changed"]

    subgraph spellDecision ["getTargetSpelling"]
        M["spellSongKeyRoot(targetKey)"] --> N{root in\nFLAT_KEY_ROOTS?}
        N -- Yes --> O["spelling: flat\nNote.fromMidi"]
        N -- No --> P["spelling: sharp\nNote.fromMidiSharps"]
    end
    G -.-> spellDecision
```

</a>

<sub>Reviews (2): Last reviewed commit: ["Merge branch &#39;main&#39; into kyujinelson/swy..."](https://github.com/justaddcl/sanbi/commit/b02b0fa7657e6c3a556722401557ba8d3f444bef) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=39863619)</sub>

### CODE REVIEW

**Comment by justaddcl** [comment]
@coderabbitai review

File: src/lib/music/chords/getTransposeSemitoneDistance.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import { type SongKey } from "@lib/constants";
+import * as Interval from "@tonaljs/interval";
+import * as Note from "@tonaljs/note";
+
+import { spellSongKeyRoot } from "./getSongKeySemitone";
+
+const modulo = (value: number, divisor: number) =>
+  ((value % divisor) + divisor) % divisor;
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/music/chords/getTransposeSemitoneDistance.ts]: Updated in c7ef03f: extracted the shared positive modulo helper into `src/lib/music/chords/modulo.ts` and imported it in both consumers.

File: src/lib/music/chords/normalizeChordAccidental.ts [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { type ChordAccidental } from "./types";
+
+export const normalizeChordAccidental = (
+  accidental: string,
+): ChordAccidental | "" | null => {
+  if (accidental === "" || accidental === undefined) {
+    return "";
+  }
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/music/chords/normalizeChordAccidental.ts:8]: Updated in c7ef03f: `normalizeChordAccidental` now accepts `string | undefined`, so the runtime guard and TypeScript signature agree.

File: src/lib/music/chords/transposeChordToken.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  const reviewReasons = getReviewReasons(parsedChordToken);
+
+  return {
+    originalToken: token,
+    transposedToken,
+    wasTransposed: transposedToken !== token,
+    status: "transposed",
+    requiresReview: reviewReasons.length > 0,
+    reviewReasons,
+  };
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/music/chords/transposeChordToken.ts:94]: Updated in c7ef03f: `wasTransposed` now requires a non-zero semitone distance, with coverage for same-key lowercase normalization so casing cleanup is not reported as pitch transposition.

File: src/lib/music/chords/__tests__/transposeChordRoot.test.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    ({ chordRoot, semitoneDistance, targetSongKey, expectedChordRoot }) => {
+      expect(
+        transposeChordRoot({
+          root: chordRoot,
+          semitoneDistance,
+          targetKey: targetSongKey,
+        }),
+      ).toBe(expectedChordRoot);
+    },
+  );
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/music/chords/__tests__/transposeChordRoot.test.ts:46]: Updated in f837acc: removed the extra row array wrapper so Jest can interpolate `$chordRoot` and `$semitoneDistance` in the parameterized test title.

File: src/lib/music/chords/__tests__/getSongKeySemitone.test.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { songKeys } from "@lib/constants";
+import { getSongKeySemitone } from "@lib/music/chords";
+
+describe("getSongKeySemitone", () => {
+  it.each(songKeys)("recognizes Sanbi song key %s", (songKey) => {
+    expect(getSongKeySemitone(songKey)).toEqual(expect.any(Number));
+  });
+});
```
[the remark this answers is not in the document]
Comment (justaddcl) [reply src/lib/music/chords/__tests__/getSongKeySemitone.test.ts:28]: Updated in f837acc: `getSongKeySemitone` now asserts the full supported key-to-semitone table instead of only checking for a number.
